src/evm-block-extractor/src/storage_clients/gcp_big_query.rs

```diff
@@ -0,0 +1,393 @@
+use std::fmt::Debug;
+
+use ethers_core::types::{Block, Transaction, TransactionReceipt, H256};
+use gcp_bigquery_client::model::dataset::Dataset;
+use gcp_bigquery_client::model::dataset_reference::DatasetReference;
+use gcp_bigquery_client::model::query_parameter::QueryParameter;
+use gcp_bigquery_client::model::query_parameter_type::QueryParameterType;
+use gcp_bigquery_client::model::query_parameter_value::QueryParameterValue;
+use gcp_bigquery_client::model::query_request::QueryRequest;
+use gcp_bigquery_client::model::table::Table;
+use gcp_bigquery_client::model::table_data_insert_all_request::TableDataInsertAllRequest;
+use gcp_bigquery_client::model::table_data_insert_all_request_rows::TableDataInsertAllRequestRows;
+use gcp_bigquery_client::model::table_field_schema::TableFieldSchema;
+use gcp_bigquery_client::model::table_schema::TableSchema;
+use gcp_bigquery_client::Client;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use super::BlockChainDB;
+use crate::constants::{BLOCKS_TABLE_ID, RECEIPTS_TABLE_ID};
+
+#[derive(Clone)]
+/// A client for BigQuery that can be used to query and insert data
+pub struct BigQueryBlockChain {
+    client: Client,
+    /// The project ID of the BigQuery project
+    project_id: String,
+    /// The dataset ID of the BigQuery table
+    ///
+    /// Can be mainnet/testnet
+    dataset_id: String,
+}
+
+impl BigQueryBlockChain {
+    /// Creates a new BigQuery client
+    pub async fn new(
+        project_id: String,
+        dataset_id: String,
+        sa_key: String,
+    ) -> anyhow::Result<Self> {
+        let service_account = yup_oauth2::parse_service_account_key(sa_key)?;
+
+        let client = Client::from_service_account_key(service_account, false).await?;
+
+        Ok(Self {
+            client,
+            project_id,
+            dataset_id,
+        })
+    }
+
+    /// Creates a new BigQuery client with a custom client
+    pub fn new_with_client(
+        project_id: String,
+        dataset_id: String,
+        client: Client,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            client,
+            project_id,
+            dataset_id,
+        })
+    }
+
+    pub async fn init(&self) -> anyhow::Result<()> {
+        let dataset = Dataset::new(&self.project_id, &self.dataset_id);
+
+        // Define tables with their respective schemas
+        let tables = [
+            (
+                BLOCKS_TABLE_ID,
+                vec![
+                    TableFieldSchema::integer("id"),
+                    TableFieldSchema::string("body"),
```

Comment:
BigQuery has a specific JSON column type, we should use it

Reply:
Changed